From 418fef7b74fb35a1c17c7e3f4ee2a94e2fc0d40e Mon Sep 17 00:00:00 2001
From: Roman Yepishev <ryepishev@google.com>
Date: Sat, 3 Jan 2026 15:02:17 -0500
Subject: [PATCH 1/3] plugins/power: logind does not emit OnExternalPower

Per https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.login1.html,
logind does not notify of OnExternalPower changes:

    @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
    readonly b OnExternalPower = ...;

So we can't use Set(). We would still like to simulate the behavior,
so update the property silently via a mock interface call.
---
 plugins/power/test.py |  7 +++++--
 tests/gsdtestcase.py  | 10 ++++++++++
 2 files changed, 15 insertions(+), 2 deletions(-)

diff --git a/plugins/power/test.py b/plugins/power/test.py
index f65b01072..155413b8b 100755
--- a/plugins/power/test.py
+++ b/plugins/power/test.py
@@ -145,8 +145,10 @@ class PowerPluginBase(gsdtestcase.GSDTestCase):
         return self.obj_session_presence_props.Get('org.gnome.SessionManager.Presence', 'status')
 
     def set_on_external_power(self, state):
-        self.logind_obj.Set('org.freedesktop.login1.Manager', 'OnExternalPower', state)
-        self.logind_obj.EmitSignal('', 'Changed', '', [], dbus_interface='org.freedesktop.DBus.Mock')
+        # logind does not send PropertiesChanged signal for OnExternalPower:
+        # So use a SetOnExternalPower method on the mock interface instead of Set().
+        logind_mock_intf = dbus.Interface(self.logind_obj, dbusmock.MOCK_IFACE)
+        logind_mock_intf.SetOnExternalPower(state)
 
         self.obj_upower.Set('org.freedesktop.UPower', 'OnBattery', not state)
         self.obj_upower.EmitSignal('', 'Changed', '', [], dbus_interface='org.freedesktop.DBus.Mock')
@@ -1036,3 +1038,4 @@ class PowerPluginTestBrightnessStep(PowerPluginBase):
 
 # avoid writing to stderr
 unittest.main(testRunner=unittest.TextTestRunner(stream=sys.stdout, verbosity=2))
+
diff --git a/tests/gsdtestcase.py b/tests/gsdtestcase.py
index 6024c6cfd..b58854d29 100644
--- a/tests/gsdtestcase.py
+++ b/tests/gsdtestcase.py
@@ -273,6 +273,16 @@ class GSDTestCase(DBusTestCase):
         self.logind_obj.AddProperty(
             'org.freedesktop.login1.Manager', 'OnExternalPower', False
         )
+        # logind does not send PropertiesChanged signal for OnExternalPower:
+        # https://www.freedesktop.org/software/systemd/man/latest/org.freedesktop.login1.html
+        #     @org.freedesktop.DBus.Property.EmitsChangedSignal("false")
+        #     readonly b OnExternalPower = ...;
+        #
+        # So use a SetOnExternalPower method on the mock interface instead of Set().
+        self.logind_obj.AddMethod(
+            dbusmock.MOCK_IFACE, 'SetOnExternalPower', 'b', '',
+            'self.props["org.freedesktop.login1.Manager"]["OnExternalPower"] = args[0]'
+        )
 
     def stop_logind(self):
         '''stop mock logind'''
-- 
GitLab


From c806b911104a32e180653545146302a25283fcda Mon Sep 17 00:00:00 2001
From: Roman Yepishev <ryepishev@google.com>
Date: Fri, 2 Jan 2026 14:08:27 -0500
Subject: [PATCH 2/3] plugins/power: Add unidle tests

This moves the existing test into a new test suite and adds coverage
for sequential AC events and WakeUpScreen handler.

Related to https://gitlab.gnome.org/GNOME/gnome-settings-daemon/-/issues/903
---
 plugins/power/meson.build | 11 ++++-
 plugins/power/test.py     | 90 ++++++++++++++++++++++++++++++++++++++-
 2 files changed, 97 insertions(+), 4 deletions(-)

diff --git a/plugins/power/meson.build b/plugins/power/meson.build
index 398710759..bbc5ee038 100644
--- a/plugins/power/meson.build
+++ b/plugins/power/meson.build
@@ -102,16 +102,23 @@ tests = [
   'BatteryLevels',
   'Brightness',
   'BrightnessStep',
+  'Unidle',
 ]
 
+timeout_by_test = {
+  # Takes very long because of SCREENSAVER_TIMEOUT_BLANK
+  'Screensaver': 180,
+  # Takes a long time because we wait for idle in multiple tests
+  'Unidle': 150
+}
+
 foreach test : tests
   test(
     'test-power @0@'.format(test),
     test_py,
     args: [ 'PowerPluginTest@0@'.format(test) ],
     env: envs,
-    # The first set of tests takes very long because of SCREENSAVER_TIMEOUT_BLANK
-    timeout: test == 'Screensaver' ? 180 : 120
+    timeout: timeout_by_test.get(test, 120)
   )
 endforeach
 
diff --git a/plugins/power/test.py b/plugins/power/test.py
index 155413b8b..a08485819 100755
--- a/plugins/power/test.py
+++ b/plugins/power/test.py
@@ -984,8 +984,12 @@ class PowerPluginTestBrightness(PowerPluginBase):
 
         self.check_suspend_no_hibernate(7)
 
-    def disabled_test_unindle_on_ac_plug(self):
-        idle_delay = round(gsdpowerconstants.MINIMUM_IDLE_DIM_DELAY / gsdpowerconstants.IDLE_DELAY_TO_IDLE_DIM_MULTIPLIER)
+class PowerPluginTestUnidle(PowerPluginBase):
+    def test_unidle_on_ac_plug(self):
+        idle_delay = round(
+            gsdpowerconstants.MINIMUM_IDLE_DIM_DELAY
+            / gsdpowerconstants.IDLE_DELAY_TO_IDLE_DIM_MULTIPLIER
+        )
         self.settings_session['idle-delay'] = idle_delay
         Gio.Settings.sync()
 
@@ -1010,6 +1014,88 @@ class PowerPluginTestBrightness(PowerPluginBase):
         # And wait a little more to see us dim again
         self.check_dim(idle_delay + 2)
 
+    def test_unidle_timer_reset_on_consecutive_ac_events(self):
+        idle_delay = (
+            round(
+                gsdpowerconstants.MINIMUM_IDLE_DIM_DELAY
+                / gsdpowerconstants.IDLE_DELAY_TO_IDLE_DIM_MULTIPLIER
+            ) + (gsdpowerconstants.POWER_UP_TIME_ON_AC * 2)
+        )
+
+        self.settings_session['idle-delay'] = idle_delay
+        Gio.Settings.sync()
+
+        # Wait for dimming on idle
+        self.check_dim(idle_delay)
+
+        # Plug in the AC
+        self.set_on_external_power(True)
+
+        # Confirm that we undimmed
+        self.check_undim(0.5)
+
+        # Sleep halfway through the timer
+        time.sleep(gsdpowerconstants.POWER_UP_TIME_ON_AC / 2)
+
+        # Unplug the AC
+        self.set_on_external_power(False)
+
+        # Check that we wait for the whole POWER_UP_TIME_ON_AC
+        self.check_no_dim(gsdpowerconstants.POWER_UP_TIME_ON_AC)
+
+        # And wait a little more to see us dim again
+        self.check_dim(gsdpowerconstants.POWER_UP_TIME_ON_AC)
+
+    def test_return_to_idle_canceled_on_activity(self):
+        idle_delay = (
+            round(
+                gsdpowerconstants.MINIMUM_IDLE_DIM_DELAY
+                / gsdpowerconstants.IDLE_DELAY_TO_IDLE_DIM_MULTIPLIER
+            )
+            + gsdpowerconstants.POWER_UP_TIME_ON_AC
+        )
+
+        self.settings_session['idle-delay'] = idle_delay
+        Gio.Settings.sync()
+
+        # Wait for idle
+        self.check_dim(idle_delay + 2)
+
+        # Plug in the AC
+        self.set_on_external_power(True)
+
+        # Check that we unblanked
+        self.check_undim(5)
+
+        # Set active state
+        self.reset_idle_timer()
+
+        # Check that we don't dim after activity
+        self.check_no_dim(gsdpowerconstants.POWER_UP_TIME_ON_AC + 1)
+
+    def test_unidle_on_screensaver_wake_up_signal(self):
+        """https://bugzilla.gnome.org/show_bug.cgi?id=726056"""
+
+        # Lock the screen via screensaver
+        self.obj_screensaver.SetActive(True)
+
+        # Check that we immediately blanked
+        self.check_blank(0.5)
+
+        # Wake up the screen
+        self.obj_screensaver.EmitSignal(
+            '', 'WakeUpScreen', '', [], dbus_interface='org.freedesktop.DBus.Mock'
+        )
+
+        # Check that we unblanked
+        self.check_unblank(5)
+
+        time.sleep(gsdpowerconstants.POWER_UP_TIME_ON_AC)
+
+        # Check that we blanked again
+        self.check_blank(1)
+
+
 class PowerPluginTestBrightnessStep(PowerPluginBase):
     def test_power_saver_on_low_battery(self):
         '''Check that the power-saver profile gets held when low on battery'''
-- 
GitLab


From 44e1cc564b02349adab38e691770f13c0e09951b Mon Sep 17 00:00:00 2001
From: Roman Yepishev <ryepishev@google.com>
Date: Sat, 27 Dec 2025 16:27:15 -0500
Subject: [PATCH 3/3] plugins/power: do not unidle for inactive session

NVIDIA drivers change the virtual terminal before suspend and restore it
upon resume.

Upon resume, there is a period of time when systemd reports that our
session is not active (we are on a different virtual terminal). On
resume, `idle_set_mode` would ignore that state change (keeping it as
"sleep"), and if `WakeUpScreen` signal from screensaver arrives at that
time, we will store "sleep" as `previous_idle_mode`, and set up a
callback to restore to that state. That causes the machine to suspend
15 seconds after resume.

This change refactors the unidle calls, refactoring common checks into
a separate function.

Up until https://gitlab.gnome.org/GNOME/mutter/-/merge_requests/3481,
mutter was resetting the idle timer on AC events and gsd-power did not
have to do anything. Once mutter starts resetting the idle timer again,
we can remove the code handling AC events.

Closes #903
---
 plugins/power/gsd-power-manager.c | 39 ++++++++++++++++++++++++-------
 1 file changed, 31 insertions(+), 8 deletions(-)

diff --git a/plugins/power/gsd-power-manager.c b/plugins/power/gsd-power-manager.c
index f0718a7f9..6c053b449 100644
--- a/plugins/power/gsd-power-manager.c
+++ b/plugins/power/gsd-power-manager.c
@@ -2279,6 +2279,35 @@ set_temporary_unidle_on_ac (GsdPowerManager *manager,
         }
 }
 
+static gboolean
+should_set_temporary_unidle_on_ac (GsdPowerManager *manager)
+{
+        /* Do not unidle if lid is closed */
+        if (manager->lid_is_closed)
+                return FALSE;
+
+        /* Do not unidle if we are not in an active session */
+        if (! manager->session_is_active)
+                return FALSE;
+
+        /* Unidle is already running, so we will need to reset the timer */
+        if (manager->temporary_unidle_on_ac_id != 0)
+                return TRUE;
+
+        /* Do not unidle if our current state isn't dim or blank */
+        if (manager->current_idle_mode != GSD_POWER_IDLE_MODE_BLANK &&
+            manager->current_idle_mode != GSD_POWER_IDLE_MODE_DIM)
+                return FALSE;
+
+        return TRUE;
+}
+
+static void
+update_temporary_unidle_on_ac (GsdPowerManager *manager)
+{
+        set_temporary_unidle_on_ac(manager, should_set_temporary_unidle_on_ac (manager));
+}
+
 static void
 up_client_on_battery_cb (UpClient *client,
                          GParamSpec *pspec,
@@ -2303,13 +2332,7 @@ up_client_on_battery_cb (UpClient *client,
 
         idle_configure (manager);
 
-        if (manager->lid_is_closed)
-                return;
-
-        if (manager->current_idle_mode == GSD_POWER_IDLE_MODE_BLANK ||
-            manager->current_idle_mode == GSD_POWER_IDLE_MODE_DIM ||
-            manager->temporary_unidle_on_ac_id != 0)
-                set_temporary_unidle_on_ac (manager, TRUE);
+        update_temporary_unidle_on_ac (manager);
 }
 
 static void
@@ -2372,7 +2395,7 @@ handle_screensaver_active (GsdPowerManager *manager,
 static void
 handle_wake_up_screen (GsdPowerManager *manager)
 {
-        set_temporary_unidle_on_ac (manager, TRUE);
+        update_temporary_unidle_on_ac (manager);
 }
 
 static void
-- 
GitLab
